### PR TITLE
Update codec version to the 4.1 version

### PR DESCRIPTION
**Pull request by DarkEld3r**
#2855 was accidentally merged (and reverted) before it was ready, so here is an updated version.

### CODE REVIEW

**Comment by bkchr** [comment]
@andresilva still has the commit in https://github.com/paritytech/substrate/pull/2801 .

File: core/finality-grandpa/src/environment.rs [hunk 1]
```diff
@@ -636,7 +636,7 @@ where
 		round: u64,
 		state: RoundState<Block::Hash, NumberFor<Block>>,
 		base: (Block::Hash, NumberFor<Block>),
-		votes: Vec<SignedMessage<Block>>,
+		votes: &HistoricalVotes<Block::Hash, NumberFor<Block>, Self::Signature, Self::Id>,
```
Comment (DemiMarie-parity) [inline core/finality-grandpa/src/environment.rs]: What is the reason for this change?

Comment (DarkEld3r) [reply core/finality-grandpa/src/environment.rs]: It is an implementation of the `Environment` trait from `finality-grandpa`. The trait definition is changed, so the implementation needs to be updated.

**Comment by andresilva** [comment]
@DarkEld3r I merged master into #2801 when master still had this change in. It's fine to merge it in this PR instead so you can keep attribution (I'll rebase and remove it from my PR once it is in). The issue is around synchronization of merges: we can't release `finality-grandpa` as is, we need to merge https://github.com/paritytech/finality-grandpa/pull/53 first since it has breaking changes, and the changes necessary to support that exist in #2801. And we can't merge this PR because it depends on an unreleased `finality-grandpa`.

File: core/finality-grandpa/src/environment.rs (hunk 1 above)
Comment (andresilva) [reply core/finality-grandpa/src/environment.rs]: https://github.com/paritytech/finality-grandpa/pull/65

**Comment by thiolliere** [comment]
also parity-codec 4.1 needs the fix of for crafted input that trigger huge allocation https://github.com/paritytech/parity-scale-codec/issues/109

**Comment by gavofyork** [comment]
@DarkEld3r is this good to be merged now?

**Comment by DarkEld3r** [comment]
@gavofyork I suppose it makes sense to wait a little bit more for `4.1.1` to be released.

**Comment by Kianenigma** [comment]
> @gavofyork I suppose it makes sense to wait a little bit more for `4.1.1` to be released ([paritytech/parity-scale-codec#114](https://github.com/paritytech/parity-scale-codec/pull/114)).

[U](https://crates.io/crates/parity-codec)pdated.

File: core/consensus/slots/Cargo.toml [hunk 2]
```diff
@@ -6,7 +6,7 @@ description = "Generic slots-based utilities for consensus"
 edition = "2018"
 
 [dependencies]
-codec = { package = "parity-codec", version = "3.2" }
+codec = { package = "parity-codec", version = "4.1" }
```
Comment (Kianenigma) [inline core/consensus/slots/Cargo.toml]: Maybe not now but: 
don't we want to put some guidelines on import names? sometimes it can be confusing. I bet you can find the crate `core/sr-primitives` along the code base with names: 
- `primitive`
- `sr-primitives`
- `substrate-primitives`
- ...

Comment (DarkEld3r) [reply core/consensus/slots/Cargo.toml]: I like the idea, though it will be hard to apply strict rules everywhere because we have a lot of similar names (for example "primitives").

**Comment by bkchr** [comment]
Can we not change the destination branch to @andresilva's grandpa branch for #2801? Merge it into this branch and merge into master with #2801 together.

**Comment by DarkEld3r** [comment]
@gavofyork I suppose it is OK to merge now.
